lib/ansible/module_utils/basic.py

```diff
@@ -1740,160 +1603,59 @@ def _check_argument_values(self, spec=None, param=None):
                 self.fail_json(msg=msg)
 
     def safe_eval(self, value, locals=None, include_exceptions=False):
-
-        # do not allow method calls to modules
-        if not isinstance(value, string_types):
-            # already templated to a datavaluestructure, perhaps?
-            if include_exceptions:
-                return (value, None)
-            return value
-        if re.search(r'\w\.\w+\(', value):
-            if include_exceptions:
-                return (value, None)
-            return value
-        # do not allow imports
-        if re.search(r'import \w+', value):
-            if include_exceptions:
-                return (value, None)
-            return value
-        try:
-            result = literal_eval(value)
-            if include_exceptions:
-                return (result, None)
-            else:
-                return result
-        except Exception as e:
-            if include_exceptions:
-                return (value, e)
-            return value
+        return safe_eval(value, locals, include_exceptions)
 
     def _check_type_str(self, value):
-        if isinstance(value, string_types):
-            return value
+        opts = {
+            'error': False,
+            'warn': False,
+            'ignore': True
+        }
 
         # Ignore, warn, or error when converting to a string.
-        # The current default is to warn. Change this in Anisble 2.12 to error.
-        common_msg = 'quote the entire value to ensure it does not change.'
-        if self._string_conversion_action == 'error':
-            msg = common_msg.capitalize()
-            raise TypeError(msg)
-        elif self._string_conversion_action == 'warn':
-            msg = ('The value {0!r} (type {0.__class__.__name__}) in a string field was converted to {1!r} (type string). '
-                   'If this does not look like what you expect, {2}').format(value, to_text(value), common_msg)
-            self.warn(msg)
-        return to_native(value, errors='surrogate_or_strict')
+        allow_conversion = opts.get(self._string_conversion_action, True)
+        try:
+            return check_type_str(value, allow_conversion)
+        except TypeError as e:
+            common_msg = 'quote the entire value to ensure it does not change.'
+            if self._string_conversion_action == 'error':
+                msg = common_msg.capitalize()
+                raise TypeError(to_native(msg))
+            elif self._string_conversion_action == 'warn':
+                msg = ('The value {0!r} (type {0.__class__.__name__}) in a string field was converted to {1!r} (type string). '
```

Comment:
Can we add a `key` name as well ? It is very confusing message if we have multiple warnings like this - 

```
[WARNING]: The value 600 (type int) in a string field was converted to u'600' (type string). If this does not look like what you expect,
quote the entire value to ensure it does not change.

 [WARNING]: The value {'Name': 'aws-blorp_rhel7'} (type dict) in a string field was converted to u"{'Name': 'aws-blorp_rhel8'}" (type
string). If this does not look like what you expect, quote the entire value to ensure it does not change.

 [WARNING]: The value 300 (type int) in a string field was converted to u'300' (type string). If this does not look like what you expect,
quote the entire value to ensure it does not change.
```

Maybe something like - 

```suggestion
                msg = ('The value {0!r} (type {0.__class__.__name__}) in a string field of argument {1} was converted to {2!r} (type string). '
```

Reply:
I'm not sure of a good way to do this since we don't have the name of the `key` available, only the value to check or convert.

Reply:
Got it. But this is very confusing error message. anyways, not a blocker as such.

Reply:
I figured the value would be enough to track down the problematic field. Even if you have multiple fields with the same value, the error only prints once (a byproduct of having the same error message text).

But I see that as a warning, it gets printed at the top of the play and that removes a lot of context. 🤔 As an `error`, it's much more clear what is happening.

There is room for improvement here, for sure.